pylint/checkers/unicode.py

```diff
@@ -0,0 +1,701 @@
+        if char.unescaped == "\r" and line.endswith(new_line):
+            ignore_pos = len(line) - 2 * byte_str_length
+        else:
+            ignore_pos = None
+
+        start: int = 0
+        pos: int = line.find(search_for, start)
+        while pos > 0:
+            if pos != ignore_pos:
+                # Calculate the
+                col = int(pos / byte_str_length)
+                result[col] = char
+            start = pos + 1
+            pos = line.find(search_for, start)
+
+    return result
+
+
+UNICODE_BOMS: Dict[str, bytes] = {
+    "utf-8": codecs.BOM_UTF8,
+    "utf-16": codecs.BOM_UTF16,
+    "utf-32": codecs.BOM_UTF32,
+    "utf-16le": codecs.BOM_UTF16_LE,
+    "utf-16be": codecs.BOM_UTF16_BE,
+    "utf-32le": codecs.BOM_UTF32_LE,
+    "utf-32be": codecs.BOM_UTF32_BE,
+}
+
+
+@lru_cache(maxsize=1)
+def get_bom_sorted() -> Dict[bytes, str]:
+    """Get all BOM sorted by size, biggest first"""
+    result = OrderedDict()
+    for codec, bom in sorted(
+        UNICODE_BOMS.items(), key=lambda x: len(x[1]), reverse=True
+    ):
+        if codec not in ("utf-16", "utf-32"):
+            # only use codecs with defined endian
+            result[bom] = codec
+    return result
+
+
+def _normalize_codec_name(codec: str) -> str:
+    """Make sure the codec name is always gives as defined in the BOM dict"""
+    codec = codec.lower()
+    return re.sub("utf[ -]?(8|16|32)[ -]?(le|be|)", r"utf-\1\2", codec)
+
+
+def _remove_bom(encoded: bytes, encoding: str) -> bytes:
+    """remove the bom if given from a line"""
+    encoding = _normalize_codec_name(encoding)
+    if not encoding.startswith("utf"):
+        return encoded
+    bom = UNICODE_BOMS[encoding]
+    if encoded.startswith(bom):
+        return encoded[len(bom) :]
+    return encoded
+
+
+def _encode_without_bom(string: str, encoding: str) -> bytes:
+    """Encode a string but remove the BOM"""
+    return _remove_bom(string.encode(encoding), encoding)
+
+
+def _byte_to_str_length(codec: str) -> int:
+    """return how many byte are usually(!) a character point"""
+    codec = codec.replace("-", "").replace(" ", "").lower()
+    if codec.startswith("utf32"):
+        return 4
+    if codec.startswith("utf16"):
+        return 2
+
+    return 1
+
+
+@lru_cache(maxsize=1000)
+def _cached_encode_search(string: str, encoding: str) -> bytes:
+    """A cached version of encode used for search pattern"""
+    return _encode_without_bom(string, encoding)
+
+
+def _fix_utf16_32_line_stream(steam: Iterable[bytes], codec: str) -> Iterable[bytes]:
+    """Handle line ending for UTF16 and UTF32 correctly
+
+    Currently Python simply strips the required zeros after \n after the
+    line ending. Leading to lines that can't be decoded propery
+    """
+    codec = _normalize_codec_name(codec)
+    if not codec.startswith("utf-16") and not codec.startswith("utf-32"):
+        yield from steam
+    else:
+        # UTF 16 and UTF-32 are sucking, so first get all the bytes in memory
```

Comment:
What are they sucking? 😄

Reply:
Because I needed to write this function to fix line endings....
The anger needed to go somewhere ;-)

Reply:
Understandable 😄